File: sdks/typescript/src/domain-error.ts
/**
 * Domain-level error mapping for Kimberlite.
 *
 * AUDIT-2026-04 S2.4 — lifts `mapKimberliteError` out of
 * `notebar/packages/kimberlite-client/src/retry.ts` into the SDK
 * so every app (React Router V7 loaders, Express handlers, CLI
 * tools) gets a single canonical translation from wire-level
 * errors to app-visible domain-error shapes.
 *
 * Typical usage in an HTTP handler:
 *
 * ```ts
 * import { asResult, DomainError } from '@kimberlite/sdk';
 *
 * export async function loader({ params }) {
 *   const result = await asResult(() => client.query('...'));
 *   if (!result.ok) return renderErrorResponse(result.err);
 *   return json(result.value);
 * }
 * ```
 */

import { KimberliteError } from './errors';

/**
 * Discriminated union of domain-facing error shapes. A 1-1 mapping
 * from wire-level `ErrorCode` to the kind of failure an app UI or
 * HTTP endpoint needs to distinguish.
 */
export type DomainError =
  | { readonly kind: 'NotFound' }
  | { readonly kind: 'Forbidden' }
  | { readonly kind: 'ConcurrentModification' }
  | { readonly kind: 'Conflict'; readonly reasons: readonly string[] }
  | { readonly kind: 'InvariantViolation'; readonly name: string }
  | { readonly kind: 'Unavailable'; readonly message: string }
  | { readonly kind: 'RateLimited' }
  | { readonly kind: 'Timeout' }
  | { readonly kind: 'Validation'; readonly message: string };

/**
 * Translate any thrown value into a `DomainError`. Non-Kimberlite
 * errors fall through to `Unavailable` with the stringified
 * message — never opaque, never reveals stack frames.
 */
export function mapKimberliteError(e: unknown): DomainError {
  if (e instanceof KimberliteError) {
    switch (e.code) {
      case 'OffsetMismatch':
        return { kind: 'ConcurrentModification' };
      case 'StreamNotFound':
      case 'TableNotFound':
      case 'TenantNotFound':
      case 'ApiKeyNotFound':
        return { kind: 'NotFound' };
      case 'AuthenticationFailed':
        return { kind: 'Forbidden' };
      case 'RateLimited':
        return { kind: 'RateLimited' };
      case 'Timeout':
        return { kind: 'Timeout' };
      case 'QueryParseError':
      case 'InvalidRequest':
      case 'InvalidOffset':
        return {
          kind: 'Validation',
          message: e.message || 'invalid request',
        };
      case 'TenantAlreadyExists':
      case 'StreamAlreadyExists':
        return {
          kind: 'Conflict',
          reasons: [e.message || String(e.code)],
        };
      default:
        return {
          kind: 'Unavailable',
          message: e.message || 'unknown error',
        };
    }
  }
  if (typeof e === 'object' && e !== null && 'message' in e) {
    const msg = (e as { message?: unknown }).message;
    return { kind: 'Unavailable', message: typeof msg === 'string' ? msg : String(e) };
  }
  return { kind: 'Unavailable', message: String(e) };
}

/** Lightweight `Result<T, E>` type so apps can avoid throwing. */
export type Result<T, E> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly err: E };

/**
 * Run `op` and translate any thrown error into a
 * `DomainError`. Returns a `Result` that HTTP handlers / React
 * Router V7 loaders can pattern-match on without a try/catch
 * around every call site.
 */
export async function asResult<T>(
  op: () => Promise<T>,
): Promise<Result<T, DomainError>> {
  try {
    return { ok: true, value: await op() };
  } catch (e) {
    return { ok: false, err: mapKimberliteError(e) };
  }
}

File: sdks/go/domain_error.go
package kimberlite

import (
	"errors"
	"strings"
)

// DomainError is a typed translation of wire-level errors into the
// categories application code actually wants to dispatch on (HTTP
// status, Slack alert, retry policy, etc.).
//
// AUDIT-2026-04 S4.2 — mirrors the TypeScript and Rust SDK
// DomainError surfaces so every downstream app that talks to
// Kimberlite from Go can do the same pattern-matching on outcomes
// without rewriting a translator.
type DomainError struct {
	// Kind classifies the failure.
	Kind DomainErrorKind
	// Message is a human-readable description.
	Message string
	// Reasons lists conflict-specific details (populated for
	// Kind=ConflictError — e.g. "stream already exists").
	Reasons []string
	// Name names the violated invariant (populated for
	// Kind=InvariantViolation).
	Name string
}

// DomainErrorKind enumerates the shapes consumers typically want to
// branch on. Deliberately coarse — fine-grained taxonomy lives in
// KimberliteError.Code.
type DomainErrorKind int

const (
	DomainKindUnavailable DomainErrorKind = iota
	DomainKindNotFound
	DomainKindForbidden
	DomainKindConcurrentModification
	DomainKindConflict
	DomainKindInvariantViolation
	DomainKindRateLimited
	DomainKindTimeout
	DomainKindValidation
)

func (e *DomainError) Error() string {
	return "kimberlite: " + e.Kind.String() + ": " + e.Message
}

// String returns a stable name suitable for logs / metrics tags.
func (k DomainErrorKind) String() string {
	switch k {
	case DomainKindNotFound:
		return "NotFound"
	case DomainKindForbidden:
		return "Forbidden"
	case DomainKindConcurrentModification:
		return "ConcurrentModification"
	case DomainKindConflict:
		return "Conflict"
	case DomainKindInvariantViolation:
		return "InvariantViolation"
	case DomainKindRateLimited:
		return "RateLimited"
	case DomainKindTimeout:
		return "Timeout"
	case DomainKindValidation:
		return "Validation"
	default:
		return "Unavailable"
	}
}

// MapKimberliteError translates any error returned by the Kimberlite
// client into a structured DomainError. Non-Kimberlite errors fall
// through to DomainKindUnavailable with the raw message — never
// opaque, never reveals stack frames.
//
//	rows, err := client.Query(...)
//	if err != nil {
//	    d := kimberlite.MapKimberliteError(err)
//	    switch d.Kind {
//	    case kimberlite.DomainKindConcurrentModification:
//	        return retry()
//	    case kimberlite.DomainKindForbidden:
//	        return http.StatusForbidden
//	    ...
//	    }
//	}
func MapKimberliteError(err error) *DomainError {
	if err == nil {
		return nil
	}

	// Sentinel errors take precedence.
	switch {
	case errors.Is(err, ErrStreamNotFound):
		return &DomainError{Kind: DomainKindNotFound, Message: err.Error()}
	case errors.Is(err, ErrPermissionDenied):
		return &DomainError{Kind: DomainKindForbidden, Message: err.Error()}
	case errors.Is(err, ErrTimeout):
		return &DomainError{Kind: DomainKindTimeout, Message: err.Error()}
	case errors.Is(err, ErrNotConnected), errors.Is(err, ErrConnectionFailed):
		return &DomainError{Kind: DomainKindUnavailable, Message: err.Error()}
	}

	// Rich KimberliteError from the wire.
	var ke *KimberliteError
	if errors.As(err, &ke) {
		return mapByCode(ke)
	}

	return &DomainError{Kind: DomainKindUnavailable, Message: err.Error()}
}

func mapByCode(e *KimberliteError) *DomainError {
	switch e.Code {
	case "OffsetMismatch":
		return &DomainError{Kind: DomainKindConcurrentModification, Message: e.Message}
	case "StreamNotFound", "TableNotFound", "TenantNotFound", "ApiKeyNotFound":
		return &DomainError{Kind: DomainKindNotFound, Message: e.Message}
	case "AuthenticationFailed":
		return &DomainError{Kind: DomainKindForbidden, Message: e.Message}
	case "RateLimited":
		return &DomainError{Kind: DomainKindRateLimited, Message: e.Message}
	case "Timeout":
		return &DomainError{Kind: DomainKindTimeout, Message: e.Message}
	case "QueryParseError", "InvalidRequest", "InvalidOffset":
		return &DomainError{Kind: DomainKindValidation, Message: e.Message}
	case "TenantAlreadyExists", "StreamAlreadyExists":
		return &DomainError{
			Kind:    DomainKindConflict,
			Message: e.Message,
			Reasons: []string{strings.TrimSpace(e.Message)},
		}
	default:
		return &DomainError{Kind: DomainKindUnavailable, Message: e.Message}
	}
}
